bubbles/commands/periodic/rule_monitoring.py

```diff
@@ -0,0 +1,356 @@
+new_subreddits: List[str] = []
+# The stack for the subreddits to process
+# The subreddits that have not been updated for the longest are at the top
+# of the stack (the end of the list)
+subreddit_stack: List[str] = []
+
+
+class SubredditRule(TypedDict):
+    # The rule number
+    index: int
+    name: str
+    description: str
+    created_time: datetime
+
+
+class RuleEntry(TypedDict):
+    last_update: datetime
+    rules: List[SubredditRule]
+
+
+# A map from a subreddit name to its rules
+# If the dict has no entry for a given sub, the rules have not been fetched yet
+# If the dict has an empty list as entry for a given sub, this sub did not define any rules
+SubredditRuleMap = Dict[str, RuleEntry]
+
+
+class RuleEdited(TypedDict):
+    """A rule has been edited."""
+
+    # The rule before the edit
+    old_rule: SubredditRule
+    # The rule after the edit
+    new_rule: SubredditRule
+
+
+class RuleChanges(TypedDict):
+    """A collection of rule changes for a sub.
+
+    The rules are compared by their index.
+    """
+
+    added: List[SubredditRule]
+    removed: List[SubredditRule]
+    edited: List[RuleEdited]
+
+
+def _save_all_rules(rules: SubredditRuleMap):
+    """Save all rules to the save file."""
+    path = RULE_MONITORING_DATA_PATH
+    dir_path = os.path.dirname(path)
+
+    # Create the data folder if it doesn't exist yet
+    if not os.path.exists(dir_path):
+        os.makedirs(dir_path, exist_ok=True)
+
+    # Write the rules to the file
+    with open(path, "w+") as file:
+        json.dump(rules, file, indent=2)
+
+
+def _save_rules_for_sub(sub_name: str, rules: Optional[List[SubredditRule]]):
+    """Save the rules for the given sub.
+
+    The changes are persisted to a file, such that it is available after restart.
+    """
+    rule_entry: RuleEntry = {
+        # It has been updated now
+        "last_updated": datetime.now(),
+        "rules": rules or [],
+    }
+
+    # Load the previous rules
+    all_rules = _load_all_rules()
+    # Add the change
+    all_rules[sub_name] = rule_entry
+    # Save it again
+    _save_all_rules(all_rules)
+
+
+def _load_all_rules() -> SubredditRuleMap:
+    """Loads all rules from the save file."""
+    path = RULE_MONITORING_DATA_PATH
+
+    if not os.path.exists(path):
+        # We don't have any data saved yet, so no rules yet
+        return {}
+
+    with open(path, "r") as file:
+        content = file.read()
+
+        if content == "" or content.isspace():
+            # There is a file, but it's empty, so no rules yet
+            return {}
+
+        # Otherwise, parse the saved rule data
+        return json.loads(content)
+
+
+def _load_rules_for_sub(sub_name) -> Optional[List[SubredditRule]]:
+    """Load the rules for the given sub."""
+    all_rules = _load_all_rules()
+
+    if rule_entry := all_rules.get(sub_name):
+        return rule_entry["rules"]
+
+    # We never checked the given sub
+    return None
```

Comment:
these two lines - why do we have it? Is this to prevent random collection of information?

Reply:
We basically need a way to detect if a subreddit is new to the queue, i.e. we have never checked the rules of that sub. In that case, we don't want to notify mods about rule changes, because we have no reference to compare the current rules to. So we just want to save the current rules to a file to compare with next time we check that sub.

The problem now is that there are also subreddits that have just no rules defined. In that case we _do_ want to notify the mods when suddenly they added a rule.

The solution: If we already checked the subreddit before and they just didn't have rules, we will have an entry for the sub in our save file, with an empty list of rules, so we return `[]` (an empty list). If we never checked the subreddit before, we won't have an entry in our save file yet and we return `None` instead.

Now, when we want to determine if any subreddit has been newly added, we can just check if `None` has been returned. Then we don't notify mods of the changes (because we have no way of knowing if there was a change), so we just save the current rules of the sub to compare to next time.